soc/arm/silabs/bg2x: support DFU

rdob-ant opened the pull request:
This PR contains contains following changes:
1. Enables flash driver for SiLabs EFR32BG22 and BG27 that was disabled by default.
2. Updates partition tables for Thunderboard EVB for those SoCs (removes scratch partition, increases size of MCUboot partition to 48K to fit recent MCUboot builds that are ~40KB), sets `zephyr,code-partition` in DTS.
3. Partition tables are moved from SoC DTS files to board DTS files, and this layout is aligned between BG22 and BG27.

With these changes, MCUboot & mcumgr can perform a DFU over UART.

rdob-ant committed:
soc/arm/silabs/bg2x: enable flash driver

It is enabled by default if Device Tree includes a flash controller,
and disabled otherwise. SoC defconfig should not touch it.

Signed-off-by: Roman Dobrodii <rdobrodii@antmicro.com>

rdob-ant committed:
dts/arm/silabs: update partition definitions

- To link image loadable by MCUboot, zephyr,code-partition
must be set in the DTS.
- Move partition definitions from SoC DTS to the board DTS.
- Remove scratch partition since MCUboot does not recommend to use it.
- Increase bootloader partitions to 48K to fit recent MCUboot.

Signed-off-by: Roman Dobrodii <rdobrodii@antmicro.com>

kmeinhar commented:
In my opinion good idea. I am using a similar patch locally in my project with EFM32PG22

Might be worth to also add `mcuboot-button0 = &button0;` to device tree
 
If I remember correctly MCUBoot uses flash erases when swapping image 0 and 1 during DFU.
Flash erases are required to erase pages (8K on BG2X).
Because of that for me MCUBoot crashed when image partitions were not page aligned.
In you partition table this is not the case for `0x45000`.

To test that you need firmware image support in the main image.
This way you can upload an image to slot 1 and mark it for testing.
MCUBoot should then crash during swap.
If this this does not crash please let me know

rdob-ant committed:
boards/arm/efr32_thunderboard: MCUboot led, button

Add mcuboot DT aliases for LED and button. These can be used by MCUboot
when recovery mode entrance via GPIO is enabled.

Signed-off-by: Roman Dobrodii <rdobrodii@antmicro.com>

rdob-ant commented:
@kmeinhar 
Indeed you're right! I overlooked the page size issue on BG22. I've re-tested it now with MCUboot and I see that although it does not crash per se, the flash driver reports erase errors due to flash areas being not aligned to the page size. I've pushed updated version of BG22 partition table and verified that with it I can upload & swap images with mcumgr several times in a row.

I've also added DT aliases in the third commit as you've proposed.

fkokosinski reviewed (approved):
LGTM
